ranger/core/loader.py

```diff
@@ -170,62 +169,39 @@ def generate(self):  # pylint: disable=too-many-branches,too-many-statements
         popenargs['stdin'] = PIPE if self.input else open(os.devnull, 'r')
         self.process = process = Popen(self.args, **popenargs)
         self.signal_emit('before', process=process, loader=self)
-        if self.input:
-            if py3:
-                import io
-                stdin = io.TextIOWrapper(process.stdin)
-            else:
-                stdin = process.stdin
-            try:
-                stdin.write(self.input)
-            except IOError as ex:
-                if ex.errno != errno.EPIPE and ex.errno != errno.EINVAL:
-                    raise
-            stdin.close()
         if self.silent and not self.read:  # pylint: disable=too-many-nested-blocks
             while process.poll() is None:
                 yield
                 if self.finished:
                     break
                 sleep(0.03)
         else:
-            selectlist = []
-            if self.read:
-                selectlist.append(process.stdout)
-            if not self.silent:
-                selectlist.append(process.stderr)
-            while process.poll() is None:
-                yield
-                if self.finished:
-                    break
-                try:
-                    robjs, _, _ = select.select(selectlist, [], [], 0.03)
-                    if robjs:
-                        robjs = robjs[0]
-                        if robjs == process.stderr:
-                            read = robjs.readline()
-                            if py3:
-                                read = safe_decode(read)
-                            if read:
-                                self.fm.notify(read, bad=True)
-                        elif robjs == process.stdout:
-                            read = robjs.read(512)
-                            if py3:
-                                read = safe_decode(read)
-                            if read:
-                                self.stdout_buffer += read
-                except select.error:
-                    sleep(0.03)
-            if not self.silent:
-                for line in process.stderr:
-                    if py3:
-                        line = safe_decode(line)
+            output = (None, None)  # (stdout, stderr)
+            try:
+                output = process.communicate(self.input)
+            except IOError as ex:
+                if ex.errno != errno.EPIPE and ex.errno != errno.EINVAL:
+                    raise
+            output = list(output)
```

Comment:
Why a list? `(stdout, stderr) = communicate()` is easier to follow imo. And I don't like inlining `safe_decode` it's much less clear what's going on now.
Also, there's a problem with communicate, it waits until the entire output is available, that's not what we want, so I regret suggesting it.

Reply:
Tuple is immutable. It must be converted to some mutable type before decoding. I did choose list.

> ... that's not what we want, so I regret suggesting it.

No problem. I will remove last commit.